File: 351004/Brazhalovich/Lab_1/src/main/java/org/example/newsapi/service/NewsService.java
package org.example.newsapi.service;


import lombok.RequiredArgsConstructor;
import org.example.newsapi.dto.request.NewsRequestTo;
import org.example.newsapi.dto.response.NewsResponseTo;
import org.example.newsapi.entity.News;
import org.example.newsapi.exception.NotFoundException;
import org.example.newsapi.mapper.NewsMapper;
import org.example.newsapi.repository.impl.NewsRepository;
import org.example.newsapi.repository.impl.UserRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class NewsService {
    private final NewsRepository newsRepository;
    private final UserRepository userRepository;
    private final NewsMapper newsMapper;

    public NewsResponseTo create(NewsRequestTo request) {
        // Валидация существования юзера
        userRepository.findById(request.getUserId())
                .orElseThrow(() -> new NotFoundException("User not found"));

        News news = newsMapper.toEntity(request);
        news.setCreated(LocalDateTime.now());
        news.setModified(LocalDateTime.now());
        return newsMapper.toDto(newsRepository.save(news));
    }

    public List<NewsResponseTo> findAll() {
        return newsRepository.findAll().stream()
                .map(newsMapper::toDto)
                .collect(Collectors.toList());
    }

    public NewsResponseTo findById(Long id) {
        return newsRepository.findById(id)
                .map(newsMapper::toDto)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + id));
    }

    public NewsResponseTo update(Long id, NewsRequestTo request) {
        News news = newsRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("News not found with id: " + id));
        newsMapper.updateEntityFromDto(request, news);
        return newsMapper.toDto(newsRepository.update(news));
    }

    public void delete(Long id) {
        if (newsRepository.findById(id).isEmpty()) {
            throw new NotFoundException("News not found with id: " + id);
        }
        newsRepository.deleteById(id);
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/service/NewsService.kt
package com.distcomp.service

import com.distcomp.dto.news.NewsRequestTo
import com.distcomp.dto.news.NewsResponseTo
import com.distcomp.entity.News
import com.distcomp.entity.User
import com.distcomp.exception.NewsNotFoundException
import com.distcomp.exception.ValidationException
import com.distcomp.mapper.NewsMapper
import com.distcomp.repository.CrudRepository
import org.springframework.stereotype.Service

@Service
class NewsService (
    val newsMapper: NewsMapper,
    val newsRepository: CrudRepository<News>,
    val userRepository: CrudRepository<User>
) {
    fun createNews(newsRequestTo: NewsRequestTo): NewsResponseTo {
        val news = newsMapper.toNewsEntity(newsRequestTo)
        newsRepository.save(news)
        val user = userRepository.findById(newsRequestTo.userId)
        news.user = user
        return newsMapper.toNewsResponse(news)
    }

    fun readNewsById(id: Long): NewsResponseTo {
        val user = newsRepository.findById(id) ?: throw NewsNotFoundException("User not found")
        return newsMapper.toNewsResponse(user)
    }

    fun readAll(): List<NewsResponseTo> {
        return newsRepository.findAll().map { newsMapper.toNewsResponse(it) }
    }

    fun updateNews(newsRequestTo: NewsRequestTo, newsId: Long?): NewsResponseTo {
        if (newsId == null || newsRepository.findById(newsId) == null) {
            throw NewsNotFoundException("User not found")
        }

        if (newsRequestTo.title.length < 2) {
            throw ValidationException("New title is too short")
        }

        val news = newsMapper.toNewsEntity(newsRequestTo)
        news.id = newsId
        newsRepository.save(news)
        val user = userRepository.findById(newsRequestTo.userId)
        news.user = user
        return newsMapper.toNewsResponse(news)
    }

    fun removeNewsById(id: Long) {
        if (newsRepository.findById(id) == null) {
            throw NewsNotFoundException("News not found")
        }

        newsRepository.removeById(id)
    }
}